Description: Reduce the size of boot.img a bit, so that we have room for a
 key modifier status check.  Thanks to Colin King.

diff -Nur -x '*.orig' -x '*~' grub2/boot/i386/pc/boot.S grub2.new/boot/i386/pc/boot.S
--- grub2/boot/i386/pc/boot.S	2010-02-09 16:59:39.000000000 +0000
+++ grub2.new/boot/i386/pc/boot.S	2010-02-09 17:00:28.000000000 +0000
@@ -27,6 +27,7 @@
 
 	/* Print message string */
 #define MSG(x)	movw $x, %si; call LOCAL(message)
+#define ERR(x)	movw $x, %si; jmp LOCAL(error_message)
 
 	.file	"boot.S"
 
@@ -233,7 +234,7 @@
 	jz	LOCAL(floppy_probe)
 
 	/* Nope, we definitely have a hard disk, and we're screwed. */
-	jmp	LOCAL(hd_probe_error)
+	ERR(hd_probe_error_string)
 
 LOCAL(final_init):
 	/* set the mode to zero */
@@ -360,22 +361,15 @@
  * BIOS Geometry translation error (past the end of the disk geometry!).
  */
 LOCAL(geometry_error):
-	MSG(geometry_error_string)
-	jmp	LOCAL(general_error)
-
-/*
- * Disk probe failure.
- */
-LOCAL(hd_probe_error):
-	MSG(hd_probe_error_string)
-	jmp	LOCAL(general_error)
+	ERR(geometry_error_string)
 
 /*
  * Read error on the disk.
  */
 LOCAL(read_error):
-	MSG(read_error_string)
-
+	movw	$read_error_string, %si
+LOCAL(error_message):
+	call	LOCAL(message)
 LOCAL(general_error):
 	MSG(general_error_string)
 
